File: src/main/java/gargoyle/xenox/util/gui/JKeyButton.java
package gargoyle.xenox.util.gui;

import javax.swing.ComponentInputMap;
import javax.swing.InputMap;
import javax.swing.JComponent;
import javax.swing.JToggleButton;
import javax.swing.SwingUtilities;
import java.awt.Toolkit;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.awt.event.KeyAdapter;
import java.awt.event.KeyEvent;
import java.awt.event.MouseAdapter;
import java.awt.event.MouseEvent;
import java.util.ArrayList;
import java.util.List;

public class JKeyButton extends JToggleButton {
    private final List<ActionListener> lateActionListeners = new ArrayList<>();
    private volatile int keyCode;

    public JKeyButton(int keyCode) {
        setFocusTraversalKeysEnabled(false);
        setInputMap(JComponent.WHEN_FOCUSED, new InputMap());
        setInputMap(JComponent.WHEN_IN_FOCUSED_WINDOW, new ComponentInputMap(this));
        setInputMap(JComponent.WHEN_ANCESTOR_OF_FOCUSED_COMPONENT, new InputMap());
        this.keyCode = keyCode;
        setText(text());
        addKeyListener(new KeyAdapter() {
            @Override
            public void keyReleased(KeyEvent e) {
                onKey(e.getKeyCode());
            }
        });
        addMouseListener(new MouseAdapter() {
            @Override
            public void mouseReleased(MouseEvent e) {
                if (SwingUtilities.isRightMouseButton(e)) {
                    onKey(0);
                }
            }
        });
    }

    public void addLateActionListener(ActionListener l) {
        lateActionListeners.add(l);
    }

    private void fireLateActionPerformed(ActionEvent event) {
        for (ActionListener listener : new ArrayList<>(lateActionListeners)) {
            listener.actionPerformed(event);
        }
    }

    public int getKeyCode() {
        return keyCode;
    }

    private void onKey(int key) {
        if (isSelected() || key == 0) {
            setSelected(false);
            keyCode = key;
            setText(text());
            fireLateActionPerformed(new ActionEvent(this, ActionEvent.ACTION_PERFORMED, null));
        }
    }

    private String text() {
        String text = KeyEvent.getKeyText(keyCode);
        return text.startsWith(Toolkit.getProperty("AWT.unknown", "Unknown")) ? "?" : text;
    }
}

File: src/main/kotlin/gargoyle/xenox/util/gui/JKeyButton.kt
package gargoyle.xenox.util.gui

import java.awt.Toolkit
import java.awt.event.ActionEvent
import java.awt.event.ActionListener
import java.awt.event.KeyAdapter
import java.awt.event.KeyEvent
import java.awt.event.MouseAdapter
import java.awt.event.MouseEvent
import javax.swing.ComponentInputMap
import javax.swing.InputMap
import javax.swing.JToggleButton
import javax.swing.SwingUtilities

class JKeyButton(keyCode: Int) : JToggleButton() {
    private val lateActionListeners: MutableList<ActionListener> = ArrayList()

    var keyCode: Int
        private set

    init {
        focusTraversalKeysEnabled = false
        setInputMap(WHEN_FOCUSED, InputMap())
        setInputMap(WHEN_IN_FOCUSED_WINDOW, ComponentInputMap(this))
        setInputMap(WHEN_ANCESTOR_OF_FOCUSED_COMPONENT, InputMap())
        this.keyCode = keyCode
        text = text()
        addKeyListener(object : KeyAdapter() {
            override fun keyReleased(e: KeyEvent) {
                onKey(e.keyCode)
            }
        })
        addMouseListener(object : MouseAdapter() {
            override fun mouseReleased(e: MouseEvent) {
                if (SwingUtilities.isRightMouseButton(e)) {
                    onKey(0)
                }
            }
        })
    }

    fun addLateActionListener(l: ActionListener) {
        lateActionListeners.add(l)
    }

    private fun fireLateActionPerformed(event: ActionEvent) {
        for (listener in ArrayList(lateActionListeners)) {
            listener.actionPerformed(event)
        }
    }

    private fun onKey(key: Int) {
        if (isSelected || key == 0) {
            isSelected = false
            keyCode = key
            text = text()
            fireLateActionPerformed(ActionEvent(this, ActionEvent.ACTION_PERFORMED, null))
        }
    }

    private fun text(): String {
        val text = KeyEvent.getKeyText(keyCode)
        return if (text.startsWith(Toolkit.getProperty("AWT.unknown", "Unknown"))) "?" else text
    }
}
